### PR TITLE
[Implement] idof<T>() for function references

### PR DESCRIPTION
**Pull request by jtenner**
All set for review. Am I missing something?

Thanks for your help again!

### CODE REVIEW

File: src/builtins.ts [hunk 1]
```diff
@@ -3592,7 +3592,12 @@ export function compileCall(
     case BuiltinSymbols.idof: {
       let type = evaluateConstantType(compiler, typeArguments, operands, reportNode);
       compiler.currentType = Type.u32;
-      if (!type) return module.unreachable();
+      if (!type || checkArgsOptional(operands, 0, 0, reportNode, compiler)) return module.unreachable();
```
Comment (dcodeIO) [inline src/builtins.ts]: The arguments check should already be covered by `evaluateConstantType` above

File: src/program.ts [hunk 2]
```diff
@@ -336,6 +336,8 @@ export class Program extends DiagnosticEmitter {
   typeClasses: Map<TypeKind,Class> = new Map();
   /** Managed classes contained in the program, by id. */
   managedClasses: Map<i32,Class> = new Map();
+  signatureTypes: Signature[] = new Array<Signature>(0);
+  signatureID: i32 = 0;
```
Comment (dcodeIO) [inline src/program.ts]: Perhaps `nextSignatureId` and move it right below `nextClassId` for consistency.

File: src/program.ts [hunk 3]
```diff
@@ -336,6 +336,8 @@ export class Program extends DiagnosticEmitter {
   typeClasses: Map<TypeKind,Class> = new Map();
   /** Managed classes contained in the program, by id. */
   managedClasses: Map<i32,Class> = new Map();
+  signatureTypes: Signature[] = new Array<Signature>(0);
```
Comment (dcodeIO) [inline src/program.ts]: Perhaps `uniqueSignatures`.

File: src/types.ts [hunk 4]
```diff
@@ -573,20 +574,37 @@ export class Signature {
...
+    this.program = program;
     this.hasRest = false;
     this.type = Type.u32.asFunction(this);
+
+    let compare: Signature;
+    let signatureTypes = program.signatureTypes;
+    let length = signatureTypes.length;
+    for (let i = 0; i < length; i++) {
+      compare = signatureTypes[i];
+      if (this.isAssignableTo(compare)) {
```
Comment (dcodeIO) [inline src/types.ts]: While this check works currently, the naming suggests that there can be multiple valid signatures. So I'm wondering if we should have an `equals` as well that we can use here, that `isAssignableTo` would be an alias of currently.

File: src/builtins.ts (hunk 1 above)
Comment (jtenner) [reply src/builtins.ts]: Okay. I'll remove the additional check.

File: src/program.ts (hunk 2 above)
Comment (jtenner) [reply src/program.ts]: Okay. Easy enough.

File: src/program.ts (hunk 3 above)
Comment (jtenner) [reply src/program.ts]: Okay.

File: src/types.ts (hunk 4 above)
Comment (jtenner) [reply src/types.ts]: Would it make sense to create an `equals<T>(value: T)` that returns `instanceof Signature && this.isAssignableTo(value)`?

Comment (dcodeIO) [reply src/types.ts]: The `equals` I am proposing is a method `Signature#equals` that is then used here, to make this line of code more clear. For instance, in the future a signature with a trailing optional argument might be considered "assignable to" a signature without that trailing optional argument, but what we want here is actual uniqueness I think, hence `equals` is a better word. So I'm proposing:

1. Rename current `Signature#isAssignableTo` to `Signature#equals`
2. Introduce a new `Signature#isAssignableTo` that calls `equals` (for now)
3. Use `isAssignableTo` where something must be "assignable to" and `equals` where something must be "equal"

Comment (jtenner) [reply src/types.ts]: Okay. That sounds reasonable. I'll work on this today.

Comment (jtenner) [reply src/types.ts]: Should be all set!

File: src/builtins.ts (hunk 1 above)
Comment (jtenner) [reply src/builtins.ts]: This was removed

File: src/program.ts (hunk 2 above)
Comment (jtenner) [reply src/program.ts]: All set.

File: src/types.ts [hunk 5]
```diff
@@ -573,20 +574,37 @@ export class Signature {
...
+  /** The program that created this signature. */
+  program: Program;
 
   /** Constructs a new signature. */
   constructor(
+    program: Program,
     parameterTypes: Type[] | null = null,
     returnType: Type | null = null,
-    thisType: Type | null = null
+    thisType: Type | null = null,
```
Comment (dcodeIO) [inline src/types.ts]: nit: trailing comma

File: src/types.ts [hunk 6]
```diff
@@ -573,20 +574,37 @@ export class Signature {
...
     this.thisType = thisType;
+    this.program = program;
     this.hasRest = false;
     this.type = Type.u32.asFunction(this);
+
+    let compare: Signature;
+    let signatureTypes = program.uniqueSignatures;
+    let length = signatureTypes.length;
+    for (let i = 0; i < length; i++) {
+      compare = signatureTypes[i];
```
Comment (dcodeIO) [inline src/types.ts]: nit: `let compare =`, and remove the var statement above

File: src/types.ts (hunk 5 above)
Comment (jtenner) [reply src/types.ts]: Ah. This is a personal style choice. Sorry about that.

File: src/types.ts (hunk 6 above)
Comment (jtenner) [reply src/types.ts]: Not a problem. This was left over from an initial implementation that required the Signature to stick around after the loop. Easy fix. Thanks for the heads up.

File: src/types.ts (hunk 5 above)
Comment (jtenner) [reply src/types.ts]: All set.

File: src/types.ts (hunk 6 above)
Comment (jtenner) [reply src/types.ts]: Should be all set.

**Comment by jtenner** [comment]
closed in favor of #739

**Pull request closed without merging**